File: src/main/java/pl/training/workshop/m5/s06_extractinterface/step3/Priceable.java
package pl.training.workshop.m5.s06_extractinterface.step3;

import java.math.BigDecimal;
import java.math.RoundingMode;

import pl.training.workshop.shared.Money;

/**
 * Krok 3 (rozwiązanie): metoda domyślna {@code vatAmount()} dodana do opublikowanego interfejsu
 * bez łamania implementacji. Używa wyłącznie operacji kontraktu, więc jest poprawna dla każdej z nich.
 * (Nowa metoda ABSTRAKCYJNA złamałaby każdą istniejącą implementację - źródłowo i binarnie.)
 */
public interface Priceable {
    Money price();

    int vatPercent();

    /** Kwota VAT zawarta w cenie brutto: brutto * stawka / (100 + stawka). */
    default Money vatAmount() {
        BigDecimal rate = BigDecimal.valueOf(vatPercent());
        return new Money(price().amount().multiply(rate)
                .divide(rate.add(BigDecimal.valueOf(100)), 2, RoundingMode.HALF_UP));
    }
}

File: typescript/src/workshop/m5/s06_extractinterface/step3/Priceable.ts
import { Decimal } from 'decimal.js';

import { Money } from '../../../shared/Money.js';

/**
 * Krok 3 (rozwiązanie): kontrakt roli bez zmian. Interfejs TS nie ma metod domyślnych, więc
 * odpowiednikiem `default Money vatAmount()` jest funkcja obok interfejsu, która przyjmuje rolę.
 * Opublikowane implementacje nie muszą nic robić, a funkcja używa wyłącznie operacji kontraktu,
 * więc jest poprawna dla każdej z nich. (Nowa WYMAGANA metoda interfejsu złamałaby kompilację
 * każdej istniejącej implementacji.)
 */
export interface Priceable {
  price(): Money;

  vatPercent(): number;
}

/** Kwota VAT zawarta w cenie brutto: brutto * stawka / (100 + stawka). */
export function vatAmount(item: Priceable): Money {
  const rate = new Decimal(item.vatPercent());
  return new Money(item.price().amount.times(rate)
    .dividedBy(rate.plus(100)).toDecimalPlaces(2, Decimal.ROUND_HALF_UP));
}
